### PR TITLE
Autonaming configuration in Configure and Check

### PR DESCRIPTION
**Pull request by mikhailshilkov**
This PR implements the bridge part of  [Autonaming Configuration](https://github.com/pulumi/pulumi/issues/1518). See [RFC](https://github.com/pulumi/pulumi/discussions/17592) for the full design.

In short, [Protobuf definition for autonaming configuration](https://github.com/pulumi/pulumi/pull/17810) introduced the protobuf changes required for the provider-side implementation. With those, a provider can:

- Declare that it supports autonaming configurations with a response flag in Configure
- Accept two extra properties in CheckRequest: a proposed name and a mode to apply it

This PR implements autonaming configuration for all bridged providers. It passes the configuration from CheckRequest all the way down to the autonaming module. The module itself is now able to understand the modes and use the proposed name appropriately:

- For `disable` mode, it will disable autonaming and throw an error if no explicit name was provided by a user
- For `enforce` mode, it will use the proposed name verbatim. Note that all checks on that name are ignored: this is by design, so that users would be able to override the provider's settings with `enforce: true` if they need to
- For `propose` mode, it will try using the proposed name but can also modify and validate it. The bridge applies transformations (e.g. lowercasing the name) and checks max length/character set requirements. See the test cases for details.

No changes are needed to the provider themselves beyond rolling to the new version of TF bridge once it's published.

The PR is split into three commits: (1) all the manual changes, (2) mechanical updates of replay tests, (3) end-to-end tests.

Resolves https://github.com/pulumi/pulumi-terraform-bridge/issues/2722

### CODE REVIEW

File: pkg/tfbridge/provider_test.go
Comment (VenelinMartinov) [inline pkg/tfbridge/provider_test.go:805]: This uses the TF SDKv1 - that is long deprecated and very few of our providers use it. Can you instead replace `shimv1` with `shimv2` here to use the TF SDKv2 for these tests?

**Review by VenelinMartinov — COMMENTED** [review]
Can we add some additional integration tests on the both the SDK and PF side?
Something like
SDKv2: https://github.com/pulumi/pulumi-terraform-bridge/blob/adb02755a88b87e503b3881c88c2a7ec1559446f/pkg/tests/schema_pulumi_test.go#L21
and 
PF: https://github.com/pulumi/pulumi-terraform-bridge/blob/adb02755a88b87e503b3881c88c2a7ec1559446f/pkg/pf/tests/schema_and_program_test.go#L32

There's a guide to writing such tests here: https://pulumi-developer-docs.readthedocs.io/projects/pulumi-terraform-bridge/en/latest/docs/guides/testing.html

File: pkg/tfbridge/info/autonaming.go
Comment (iwahbe) [inline pkg/tfbridge/info/autonaming.go:174]: Will auto-naming always be set on new engines?

Shouldn't `ModePropose` be the same as our prior default, or is it different.

File: pkg/tfbridge/info/info.go
Comment (iwahbe) [inline pkg/tfbridge/info/info.go:680]: I don't understand what the different between these 2 options is from the provider's perspective.

Can you provide an example where the user sees different behavior with `ModeEnforce` is set vs `ModeDisable`?

File: pkg/tfbridge/info/info.go
Comment (iwahbe) [inline pkg/tfbridge/info/info.go:675]: Let's avoid taking the very generic namespace `Mode` for these enums. 

```suggestion
	ComputeDefaultAutonamingModePropose ComputeDefaultAutonamingOptionsMode = 0
```

Also, can we use `iota` here, or is there a reason why not (if not, we need a comment explaining why not).

File: pkg/tfbridge/info/autonaming.go
Comment (mikhailshilkov) [reply pkg/tfbridge/info/autonaming.go:174]: No, it will only be set if users provide an explicit autonaming configuration. It will stay nil for everyone else (majority of users).

**Review by iwahbe — COMMENTED** [review]
I'd like to see replay tests before merging, so I can see what gets sent across the wire.

I don't see where functions like `hex(4)` get implemented here, so I guess those are implemented engine-side.

File: pkg/tfbridge/info/info.go
Comment (mikhailshilkov) [reply pkg/tfbridge/info/info.go:680]: Enforce will come with a value of proposed name and the provider will use that proposed name to populate the name value if it’s not specified by the user.

Disable will never populate the name if it’s not provided by the user.

**Comment by t0yv0** [comment]
Maybe that ship has sailed.

File: pkg/tfbridge/info/autonaming.go
Comment (t0yv0) [inline pkg/tfbridge/info/autonaming.go:210]: It's interesting - I feel like I should be able to answer this, but I forget; will this surface to the user as a gRPC failure trace, or will it surface as a nicely formatted error, in particular will it be associated with the resource and field? I suspect we'd want the nice formatting, but I'd need some digging to find what's necessary to make it happen.

**Comment by t0yv0** [comment]
Another curious thing to think through is state interactions. Bridge autonaming is stateful I believe and will preserve the value of a property from state instead of computing a new value, if such a value is present. I think that is a reasonable default, but what happens with this new functionality? Is ignoring users's options in that case the expected behavior or should we detect and print a warning in such cases? Perhaps worth thinking through!

**Comment by t0yv0** [comment]
Left a few comments but LGTM mostly, this looks like a great much requested feature!

File: pkg/tfbridge/info/autonaming.go
Comment (mikhailshilkov) [reply pkg/tfbridge/info/autonaming.go:210]: This is the experience:

```
     Type                         Name          Plan       Info
 +   pulumi:pulumi:Stack          azure-ts-dev  create     
 +   ├─ azure:core:ResourceGroup  rg            create     
 +   ├─ azure:storage:Account     MySa          create     
     └─ azure:storage:Container   MyContainer              1 error

Diagnostics:
  azure:storage:Container (MyContainer):
    error: automatic naming is disabled but no explicit name was provided
```

**Comment by mikhailshilkov** [comment]
> Perhaps a pedantic nit but worth calling out that this will not apply to first-class providers and data sources; I think the AutoName machinery in the bridge is wired to run in all those cases but likely not really used much for this; the intent is definitely to use it for resources.

I think that's fine? I suspect first-class providers won't have any autonamed properties, and I'm not even sure what autonaming can mean for data sources, but Check won't be called, so this PR will have no effect.

> Another curious thing to think through is state interactions. Bridge autonaming is stateful I believe and will preserve the value of a property from state instead of computing a new value, if such a value is present. I think that is a reasonable default, but what happens with this new functionality? Is ignoring users's options in that case the expected behavior or should we detect and print a warning in such cases? Perhaps worth thinking through!

Yes, the same will hold: autonaming only applies at the time of resource creation. The [RFC](https://github.com/pulumi/pulumi/discussions/17592) spells this out:

> Changing the autonaming setting on an existing stack doesn’t cause any immediate changes. It will only affect any newly created resources on this stack, including resources being replaced for unrelated reasons. To re-create resources with new names, e.g. on a dev stack, the user would need to destroy the stack and update it again.

**Comment by mikhailshilkov** [comment]
This PR is mostly ready accept for a single TODO that requires another pu/pu release.

 > Can we add some additional integration tests on the both the SDK and PF side?

@VenelinMartinov Great hint, added those now.

> I'd like to see replay tests before merging

@iwahbe Venelin pointed me towards integration tests instead of replay tests. Here is a sample gRPC log from my manual test:

```
{"method":"/pulumirpc.ResourceProvider/Configure","request":{"variables":{"azure:config:location":"WestEurope","azure:config:subscriptionId":"0282681f-7a9e-424b-80b2-96babd57a8a1"},"args":{"location":"WestEurope","subscriptionId":{"4dabf18193072939515e22adb298388d":"1b47061264138c4ac30d75fd1eb44270","value":"0282681f-7a9e-424b-80b2-96babd57a8a1"},"version":"6.10.0"},"acceptSecrets":true,"acceptResources":true,"sendsOldInputs":true,"sendsOldInputsToDelete":true},"response":{"supportsPreview":true,"supportsAutonamingConfiguration":true},"metadata":{"kind":"resource","mode":"client","name":"azure"}}
{"method":"/pulumirpc.ResourceProvider/Check","request":{"urn":"urn:pulumi:dev::azure-ts::azure:core/resourceGroup:ResourceGroup::rg","olds":{},"news":{},"randomSeed":"Q2m36uMCs9Y7Rqbz9zr3Ji+GVtKQ/SEJriFHlhBc65c=","name":"rg","type":"azure:core/resourceGroup:ResourceGroup","autonaming":{"proposedName":"mikhail-pulumi-corp-azure-ts-dev-rg-sn16"}},"response":{"inputs":{"__defaults":["location","name"],"location":"WestEurope","name":"mikhail-pulumi-corp-azure-ts-dev-rg-sn16"}},"metadata":{"kind":"resource","mode":"client","name":"azure"}}
{"method":"/pulumirpc.ResourceProvider/Check","request":{"urn":"urn:pulumi:dev::azure-ts::azure:storage/account:Account::MySa","olds":{},"news":{"accountReplicationType":"LRS","accountTier":"Standard","resourceGroupName":"mikhail-pulumi-corp-azure-ts-dev-rg-sn16"},"randomSeed":"WTp/OGn3oXy9fljVlPxQe4KTk4F5tJHeIpyQiXkQyqY=","name":"MySa","type":"azure:storage/account:Account","autonaming":{"proposedName":"barMySa23vc"}},"response":{"inputs":{"__defaults":["accountKind","allowNestedItemsToBePublic","crossTenantReplicationEnabled","defaultToOauthAuthentication","dnsEndpointType","httpsTrafficOnlyEnabled","infrastructureEncryptionEnabled","isHnsEnabled","localUserEnabled","location","minTlsVersion","name","nfsv3Enabled","publicNetworkAccessEnabled","queueEncryptionKeyType","sftpEnabled","sharedAccessKeyEnabled","tableEncryptionKeyType"],"accountKind":"StorageV2","accountReplicationType":"LRS","accountTier":"Standard","allowNestedItemsToBePublic":true,"crossTenantReplicationEnabled":false,"defaultToOauthAuthentication":false,"dnsEndpointType":"Standard","httpsTrafficOnlyEnabled":true,"infrastructureEncryptionEnabled":false,"isHnsEnabled":false,"localUserEnabled":true,"location":"04da6b54-80e4-46f7-96ec-b56ff0331ba9","minTlsVersion":"TLS1_2","name":"barmysa23vc","nfsv3Enabled":false,"publicNetworkAccessEnabled":true,"queueEncryptionKeyType":"Service","resourceGroupName":"mikhail-pulumi-corp-azure-ts-dev-rg-sn16","sftpEnabled":false,"sharedAccessKeyEnabled":true,"tableEncryptionKeyType":"Service"}},"metadata":{"kind":"resource","mode":"client","name":"azure"}}
{"method":"/pulumirpc.ResourceProvider/Check","request":{"urn":"urn:pulumi:dev::azure-ts::azure:storage/container:Container::MyContainer","olds":{},"news":{"containerAccessType":"private","name":"mycontainer","storageAccountName":"barmysa23vc"},"randomSeed":"6rY5jOUCxzl4h9RGwROEtvSMmlUVxKjDRVuk807R+bc=","name":"MyContainer","type":"azure:storage/container:Container","autonaming":{"mode":"DISABLE"}},"response":{"inputs":{"__defaults":["encryptionScopeOverrideEnabled"],"containerAccessType":"private","encryptionScopeOverrideEnabled":true,"name":"mycontainer","storageAccountName":"barmysa23vc"}},"metadata":{"kind":"resource","mode":"client","name":"azure"}}
```

**Review by VenelinMartinov — APPROVED** [review]
LGTM, thanks for adding the tests. Is the plan to wait for the next pu/pu release before merging this in?

**Review by t0yv0 — APPROVED** [review]

**Pull request merged**
